### ISSUE TITLE
Rule proposal: functions should not be `async` unless they `await`

### ISSUE THREAD

**Issue by benmccann** [typescript-eslint/typescript-eslint#9284]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+new+plugin+rule%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] My proposal specifically checks TypeScript syntax, or it proposes a check that requires type information to be accurate.
- [X] My proposal is not a "formatting rule"; meaning it does not just enforce how code is formatted (whitespace, brace placement, etc).
- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Description

The rule would check that we don't cause unnecessary performance losses. The `async` keyword should not be used where TypeScript can already determine that a `Promise` is being returned and there is no `await` as adding `async` in that case will only cause a slight performance hit

### Fail Cases

```TypeScript
/**
 * @param {RegExp} re
 * @param {(...match: any[]) => Promise<MappedCode>} get_replacement
 * @param {string} source
 */
async function calculate_replacements(re, get_replacement, source) {
	/**
	 * @type {Array<Promise<import('./private.js').Replacement>>}
	 */
	const replacements = [];
	source.replace(re, (...match) => {
		replacements.push(
			get_replacement(...match).then((replacement) => {
				const matched_string = match[0];
				const offset = match[match.length - 2];
				return { offset, length: matched_string.length, replacement };
			})
		);
		return '';
	});
	return Promise.all(replacements);
}
```

### Pass Cases

From https://github.com/sveltejs/svelte/blob/862949d22abf2996594b4315c04fc97e18bc4408/packages/svelte/src/compiler/preprocess/replace_in_code.js#L23:

```TypeScript
/**
 * @param {RegExp} re
 * @param {(...match: any[]) => Promise<MappedCode>} get_replacement
 * @param {string} source
 */
function calculate_replacements(re, get_replacement, source) {
	/**
	 * @type {Array<Promise<import('./private.js').Replacement>>}
	 */
	const replacements = [];
	source.replace(re, (...match) => {
		replacements.push(
			get_replacement(...match).then((replacement) => {
				const matched_string = match[0];
				const offset = match[match.length - 2];
				return { offset, length: matched_string.length, replacement };
			})
		);
		return '';
	});
	return Promise.all(replacements);
}
```

### Additional Info

There is https://typescript-eslint.io/rules/promise-function-async, but it's almost exactly opposite of this. I recently turned on a handful of promise-related rules in `sveltejs/svelte`, but got a ton of pushback on `@typescript-eslint/promise-function-async`. The consensus seemed to be that the rule seemed like bad practice. The description of the rule in the docs doesn't seem to match at all what it actually does. I think either its docs could be improved or that rule could be removed

**Comment by kirkwaiblinger** [comment]
Hi @benmccann!

- we have [require-await](https://typescript-eslint.io/rules/require-await/), which bans `async`-without-await, _unless_ a promise is returned.
- if you _really_ want to ban `async` without `await` regardless of promise returns, technically, you could just use the [base rule](https://eslint.org/docs/latest/rules/require-await) rather than the typescript-eslint version. That rule exists and does what you are describing, though I personally think it is a bad idea to do so.
- Coincidentally, i was looking at the docs page for promise-function-async a few minutes ago, and had the exact same thought, that it sorely needs to be updated. Maybe there's an open issue for that already, haven't checked yet.

**Comment by Josh-Cena** [comment]
> The consensus seemed to be that the rule seemed like bad practice.

Do you have specific examples in mind? I agree that the docs are not good and the rule has _a lot_ of edge cases (I was recently working on this and... I don't really want to think deeply about all the edge cases that it would have failed on). However I think the rule generally is a good idea since it prevents Zalgo.

**Comment by benmccann** [comment]
> you could just use the [base rule](https://eslint.org/docs/latest/rules/require-await) rather than the typescript-eslint version. That rule exists and does what you are describing, though I personally recommend against it.

It doesn't quite do what I'm describing, unfortunately. It doesn't check that a `Promise` would be returned without the `async`. That's why it would need to live in `typescript-eslint` rather than base. E.g. it triggers on the following in the Svelte codebase:

```
/**
 * @template V
 * @param {V} value
 * @param {() => Promise<V>} fallback lazy because could contain side effects
 * @returns {Promise<V>}
 */
export async function value_or_fallback_async(value, fallback) {
	return value === undefined ? fallback() : value;
}
```

However, if you get rid of the `async` then a `Promise` is no longer being returned (it may return `value`, which isn't necessarily a `Promise`).

The rule I'm describing shouldn't trigger on such a function. If you remove the `async` above it changes the behavior of the function. I'm suggesting to remove it only when it adds unnecessary performance overhead and doesn't offer any value in terms of behavior.

> Do you have specific examples in mind?

Yes, the one in the issue description is the one we examined most closely. There's a link there to the Svelte codebase in case you want to see the code in context.

If it's helpful to have more examples, I believe that for all of the places it suggested adding `async` the feeling was that it would not provide value though I don't know if the others were examined quite as closely:
```
│ packages/svelte/src/compiler/preprocess/replace_in_code.js
│   23:1  error  Functions that return promises must be async  @typescript-eslint/promise-function-async
│ packages/svelte/src/motion/spring.js
│    80:2  error  Functions that return promises must be async  @typescript-eslint/promise-function-async
│   131:3  error  Functions that return promises must be async  @typescript-eslint/promise-function-async
│ packages/svelte/src/motion/tweened.js
│    91:2  error  Functions that return promises must be async  @typescript-eslint/promise-function-async
│   148:3  error  Functions that return promises must be async  @typescript-eslint/promise-function-async
```

- https://github.com/sveltejs/svelte/blob/862949d22abf2996594b4315c04fc97e18bc4408/packages/svelte/src/compiler/preprocess/replace_in_code.js#L23
- https://github.com/sveltejs/svelte/blob/862949d22abf2996594b4315c04fc97e18bc4408/packages/svelte/src/motion/spring.js#L80
- https://github.com/sveltejs/svelte/blob/862949d22abf2996594b4315c04fc97e18bc4408/packages/svelte/src/motion/spring.js#L131
- https://github.com/sveltejs/svelte/blob/862949d22abf2996594b4315c04fc97e18bc4408/packages/svelte/src/motion/tweened.js#L91
- https://github.com/sveltejs/svelte/blob/862949d22abf2996594b4315c04fc97e18bc4408/packages/svelte/src/motion/tweened.js#L148

> However I think the rule generally is a good idea since it prevents Zalgo.

I'm not sure what that means.

**Comment by kirkwaiblinger** [comment]
> > you could just use the [base rule](https://eslint.org/docs/latest/rules/require-await) rather than the typescript-eslint version. That rule exists and does what you are describing, though I personally recommend against it.
> 
> It doesn't quite do what I'm describing, unfortunately. It doesn't check that a `Promise` would be returned without the `async`. That's why it would need to live in `typescript-eslint` rather than base. 

Ah, ok, thanks for clarifying! I misunderstood the details there.

> I'm suggesting to remove it only when it adds unnecessary performance overhead and doesn't offer any value in terms of behavior.

I'm not familiar with the performance implication. Are you aware of a reference for that? There _is_ a slight benefit behaviorally for the rejection case, though, see

```ts
function promiseFunction() {
   if (Math.random() < 0.3) {
       throw new Error('thrown error');
   } else if (Math.random() < 0.5) {
       return Promise.resolve('resolved value');
   } else {
       return Promise.reject('rejected error');
   }
}

async function asyncFunction() {
   if (Math.random() < 0.3) {
       throw new Error('thrown error');
   } else if (Math.random() < 0.5) {
       return Promise.resolve('resolved value');
   } else {
       return Promise.reject('rejected error');
   }
}

function callAsyncFunction() {
   // this will never throw an exception. 
   // All errors will be logged to the console in the catch handler.
   asyncFunction().then(console.log, console.error);
   
   // a third of the time exceptions will be synchronously caught
   // a third of the time an exception will be logged in the rejection handler.
   try {
       promiseFunction().catch(console.log, console.error);
   } catch (e) {
       console.error(e);
   }
}
```

**Comment by kirkwaiblinger** [comment]
I am also curious, under the premise that the `async`-without-`await` is harmful, wouldn't that imply that you'd anyways want to fix
```ts
async function promiseOrValueFunction(): Promise<number> {
    if (Math.random() < 0.5) {
        return 3;
    } else {
        return Promise.resolve(4);
    }
}
```
to 
```ts
function promiseOrValueFunction(): Promise<number> {
    if (Math.random() < 0.5) {
        return Promise.resolve(3);
    } else {
        return Promise.resolve(4);
    }
}
```
rather than let the linter allow that case? (which, you could use the base rule to achieve. You'd have to manually wrap the value returns, but an explicit return type would solve that problem for you)

**Comment by kirkwaiblinger** [comment]
> > However I think the rule generally is a good idea since it prevents Zalgo.
> 
> I'm not sure what that means.

I am also curious to know what this means lol 😆

**Comment by benmccann** [comment]
> I'm not familiar with the performance implication. Are you aware of a reference for that?

I can't find a reference for it, but three separate Svelte maintainers who are all better versed in JS than myself agreed. Some excerpts from conversations about it below.

this will result in four promises being created rather than two:
```
async function get(thing) {
  return fetch(`${base}/${thing}`).then(async (r) => r.json());
}
```

and the evidence is where it gets shotty, cuz it's V8 source and as soon as you start relying on V8 behavior/characteristics it's "read the source" or "wait for v8 blog post"
but the overhead is that it's creating new microtasks, states, and all the function handler plumbing to make sure tasks are queued, executed, and torn down. still all has to happen even if the promise immediately returns and/or does 

> I am also curious, under the premise that the async-without-await is harmful, wouldn't that imply that...
> which, you could use the base rule to achieve

Ah, maybe the base rule is preferable here actually to what I was proposing. Though that also makes me wonder if the `@typescript-eslint/require-await` behavior should just be that of the base rule by default

**Comment by kirkwaiblinger** [comment]
> I can't find a reference for it, but three separate Svelte maintainers who are all better versed in JS than myself agreed. Some excerpts from conversations about it below.
> 
> this will result in four promises being created rather than two:
> 
> ```
> async function get(thing) {
>   return fetch(`${base}/${thing}`).then(async (r) => r.json());
> }
> ```
> 
> and the evidence is where it gets shotty, cuz it's V8 source and as soon as you start relying on V8 behavior/characteristics it's "read the source" or "wait for v8 blog post" but the overhead is that it's creating new microtasks, states, and all the function handler plumbing to make sure tasks are queued, executed, and torn down. still all has to happen even if the promise immediately returns and/or does

Yeah, the reason I mention a reference is that claims around performance remind me of this, related issue for one of the eslint base rules, https://github.com/eslint/eslint/issues/18166, which is... controversial at best 🙂. And my guess is that that would be relatively unlikely to be a deciding factor in rule logic unless there were a pretty extreme difference.

> Ah, maybe the base rule is preferable here actually to what I was proposing. 

Hope this has been helpful in addressing your request! 🙂 

> Though that also makes me wonder if the `@typescript-eslint/require-await` behavior should just be that of the base rule by default

Eh, even if we granted for the sake of argument that the behavior of the base rule were preferable, the extension uses type information, so what you'd have then is just a significantly-harder-to-configure, worse version of the base rule. Probably would make more sense to just document it as "when not to use" and remark that some find the base rule preferable.

**Comment by benmccann** [comment]
> Hope this has been helpful in addressing your request! 🙂

yes, thank you!!

> the extension uses type information

In general that's super valuable, but for this particular rule I'm not sure the type information can be used in a way that results in a more valuable rule

> what you'd have then is just a significantly-harder-to-configure, worse version of the base rule

Why would it be harder to configure? The version in this repo uses the options from the base rule, so it seems like the configuration is the same in either case

**Comment by kirkwaiblinger** [comment]
> Why would it be harder to configure? 

I just mean, configuring typed linting in the first place is a hassle, since you have to supply the tsconfig and all that. It's much easier to get through the one-time-setup for an eslint core rule or non-type-aware rule. Though, of course, if you have any type-aware rules already successfully running, the marginal cost is zero for additional ones.

**Comment by kirkwaiblinger** [comment]
not to mention potential runtime costs. Short version - if a rule _can_ be written without type information, we significantly prefer to do so.

**Comment by bradzacher** [comment]
From the perf side of things - people very much overstate the impact.

Removing the `async` is a nano-optimisation - you're really talking about shaving off a few nanoseconds from an invocation. As in you need to be doing such operations a million times before you impact perf by a millisecond.

I fully understand the style and the want to be as efficient as possible! But it's not _really_ a perf concern.

**Comment by benmccann** [comment]
Ah, I think there was some confusion because when I said "though that also makes me wonder if the @typescript-eslint/require-await behavior should just be that of the base rule by default" then I was really asking if we even need `@typescript-eslint/require-await` at all

But yeah, I think the typescript version actually is better here in most cases as it's less restrictive in how the code is written. In the Svelte codebase there's very little difference between the two rules so I didn't appreciate that initially, but I tried applying the base rule in another project and some places it was a little annoying to have to switch the code in places where performance doesn't matter like tests. And of course even in the main code it's admittedly a micro-optimization. So thanks for talking through this and I agree with the conclusion of closing it

**Comment by Josh-Cena** [comment]
> I'm not sure what that means.

This is what I mean: https://developer.mozilla.org/en-US/docs/Web/JavaScript/Guide/Using_promises#timing, https://blog.izs.me/2013/08/designing-apis-for-asynchrony/. You should not have a function that's "sometimes synchronous, sometimes asynchronous" because that makes it hard to orchestrate task order.

> ```js
> async function get(thing) {
>   return fetch(`${base}/${thing}`).then(async (r) => r.json());
> }
> ```

Why not

```js
async function get(thing) {
  const res = await fetch(`${base}/${thing}`);
  return await r.json();
}
```

? Also you might be overthinking about the impact of rewrapping promises—engines are very good at optimizing that away.

**Comment by bradzacher** [comment]
Just cos I was curious. [A perf comparison of having the async](https://perf.link/#eyJpZCI6ImxxZjY5YnZsM3AwIiwidGl0bGUiOiJVbm5lY2Vzc2FyeSBhc3luYyIsImJlZm9yZSI6ImNvbnN0IHNsZWVwID0gKCkgPT4gUHJvbWlzZS5yZXNvbHZlKCk7IiwidGVzdHMiOlt7Im5hbWUiOiJObyBBc3luYyIsImNvZGUiOiJmdW5jdGlvbiBmbigpIHtcbiAgcmV0dXJuIHNsZWVwKCk7XG59XG5mbigpLnRoZW4oKCkgPT4gY29uc29sZS5sb2coKSk7IiwicnVucyI6WzI3OTgwMDAsNDYxODAwMCwxNzgwMDAwLDY4MzEwMDAsNDExMDAwLDQwOTcwMDAsOTYwMDAwLDQxMDAwMCwxODAwMCw2ODMxMDAwLDQxMTAwMCw2ODI2MDAwLDUzMTIwMDAsMjAwMCwyMTYwMDAsMjQxNzAwMCwzMzgxMDAwLDEyODIwMDAsMjI4MDAwLDE2MjkwMDAsMjI1NDAwMCw1NTM5MDAwLDEzNTUwMDAsNTk4NjAwMCw1NTc2MDAwLDMzNDUwMDAsNjgzMTAwMCwzNTA3MDAwLDIxNzEwMDAsNDM2MjAwMCwzNzEwMDAwLDQyMTUwMDAsNDExMDAwLDEyODEwMDAsMTU3NjAwMCw1NjYwMDAwLDQxMTAwMCw1NzY4MDAwLDY4MzEwMDAsNjgzMTAwMCw0ODcyMDAwLDE4NDYwMDAsMzA5MDAwLDIzOTAwMDAsMTA0NzAwMCwxOTU2MDAwLDEwNzYwMDAsNTIwOTAwMCwxMzY4MDAwLDEwMDAsNTIxMjAwMCw2ODMxMDAwLDEzMzYwMDAsNjI5MDAwLDg2OTAwMCw0NjQwMDAwLDExMzQwMDAsNDExMDAwLDU1MzQwMDAsNjM2MTAwMCw0NDYyMDAwLDM2MzYwMDAsNjgzMTAwMCw2MjQ0MDAwLDY4MzEwMDAsMzY1MTAwMCw2NzQzMDAwLDQzNDYwMDAsNDExMDAwLDI0ODYwMDAsNDAzMDAwLDM5MzMwMDAsNjgzMTAwMCw2ODMxMDAwLDE3NzYwMDAsNTQ1NjAwMCw2NzE0MDAwLDUyNDEwMDAsNDYyMDAwLDY4MzEwMDAsNjY0ODAwMCwyODI3MDAwLDE0OTEwMDAsNjgzMTAwMCw0NDQwMDAsMjMyNjAwMCw0MjU4MDAwLDEzODAwMCw2NzQ0MDAwLDE5NzcwMDAsNTEyMDAwLDQ3MDQwMDAsMTA5MzAwMCwxOTIyMDAwLDg4NjAwMCw2ODMxMDAwLDQxMTAwMCwzMzMyMDAwLDM2NjcwMDAsMTQwMzAwMF0sIm9wcyI6MzI5NjczMH0seyJuYW1lIjoiQXN5bmMiLCJjb2RlIjoiYXN5bmMgZnVuY3Rpb24gZm4oKSB7XG4gIHJldHVybiBzbGVlcCgpO1xufVxuZm4oKS50aGVuKCgpID0%2BIGNvbnNvbGUubG9nKCkpOyIsInJ1bnMiOlsyMDM3MDAwLDI0NzAwMDAsNzAwMDAwLDM1NjkwMDAsNDMxMDAwLDI1NjEwMDAsMTIyMTAwMCwzNTY5MDAwLDM1NjkwMDAsMzU2OTAwMCw1MTEwMDAsMjIxMTAwMCwzNTY5MDAwLDEwMDAsMzU2OTAwMCw0MDkwMDAsMTI2OTAwMCw1NzgwMDAsMTQxMDAwLDEwMDAwLDEyMDYwMDAsMTgxMzAwMCw2ODEwMDAsMzU2OTAwMCwzNTY5MDAwLDIyMjUwMDAsMzU2OTAwMCwxNTI0MDAwLDIzNjMwMDAsMTU5MDAwMCw1MTEwMDAsMjc0MTAwMCwzOTAwMDAsOTk5MDAwLDUxMTAwMCwzNTY5MDAwLDM1NjkwMDAsMjgzMDAwMCwzNTY5MDAwLDIzNTYwMDAsMzA1NDAwMCwxMzQ4MDAwLDM1MTAwMCwxNjkxMDAwLDUxMTAwMCw4OTYwMDAsNjUwMDAsMzU2OTAwMCwxMDAxMDAwLDM1NjkwMDAsMzU2OTAwMCwzNTY5MDAwLDkxMzAwMCwxODE2MDAwLDUxMTAwMCwyNDM3MDAwLDg4MjAwMCw2NjgwMDAsMzU2OTAwMCwzNTY5MDAwLDM0MzEwMDAsMTgxMDAwMCwzNTY5MDAwLDM1NjkwMDAsMzU2OTAwMCwyNTkyMDAwLDM1NjkwMDAsMzE3MzAwMCw1MDIwMDAsMjMwMTAwMCwzMDA0MDAwLDIwNzIwMDAsMzU2OTAwMCwzNTY5MDAwLDEzMTEwMDAsMzU2OTAwMCwzNTY5MDAwLDE5MzUwMDAsNDI1MDAwLDM1NjkwMDAsMzAwMjAwMCwxODAxMDAwLDEwMzUwMDAsMzU2OTAwMCw5MDAwLDE2MjEwMDAsMjQwODAwMCwxMDMwMDAsMzM0MDAwMCwxNTYyMDAwLDUxMTAwMCwyNjUyMDAwLDY4MjAwMCwxMzU2MDAwLDg1MDAwMCwzNTY5MDAwLDUxMTAwMCwyMTgzMDAwLDI5MTMwMDAsMTc5OTAwMF0sIm9wcyI6MjA2ODI5MH0seyJuYW1lIjoiQXN5bmMgRG93bmxldmVsbGVkIiwiY29kZSI6InZhciBfX2F3YWl0ZXIgPSAodGhpcyAmJiB0aGlzLl9fYXdhaXRlcikgfHwgZnVuY3Rpb24gKHRoaXNBcmcsIF9hcmd1bWVudHMsIFAsIGdlbmVyYXRvcikge1xuICAgIGZ1bmN0aW9uIGFkb3B0KHZhbHVlKSB7IHJldHVybiB2YWx1ZSBpbnN0YW5jZW9mIFAgPyB2YWx1ZSA6IG5ldyBQKGZ1bmN0aW9uIChyZXNvbHZlKSB7IHJlc29sdmUodmFsdWUpOyB9KTsgfVxuICAgIHJldHVybiBuZXcgKFAgfHwgKFAgPSBQcm9taXNlKSkoZnVuY3Rpb24gKHJlc29sdmUsIHJlamVjdCkge1xuICAgICAgICBmdW5jdGlvbiBmdWxmaWxsZWQodmFsdWUpIHsgdHJ5IHsgc3RlcChnZW5lcmF0b3IubmV4dCh2YWx1ZSkpOyB9IGNhdGNoIChlKSB7IHJlamVjdChlKTsgfSB9XG4gICAgICAgIGZ1bmN0aW9uIHJlamVjdGVkKHZhbHVlKSB7IHRyeSB7IHN0ZXAoZ2VuZXJhdG9yW1widGhyb3dcIl0odmFsdWUpKTsgfSBjYXRjaCAoZSkgeyByZWplY3QoZSk7IH0gfVxuICAgICAgICBmdW5jdGlvbiBzdGVwKHJlc3VsdCkgeyByZXN1bHQuZG9uZSA%2FIHJlc29sdmUocmVzdWx0LnZhbHVlKSA6IGFkb3B0KHJlc3VsdC52YWx1ZSkudGhlbihmdWxmaWxsZWQsIHJlamVjdGVkKTsgfVxuICAgICAgICBzdGVwKChnZW5lcmF0b3IgPSBnZW5lcmF0b3IuYXBwbHkodGhpc0FyZywgX2FyZ3VtZW50cyB8fCBbXSkpLm5leHQoKSk7XG4gICAgfSk7XG59O1xudmFyIF9fZ2VuZXJhdG9yID0gKHRoaXMgJiYgdGhpcy5fX2dlbmVyYXRvcikgfHwgZnVuY3Rpb24gKHRoaXNBcmcsIGJvZHkpIHtcbiAgICB2YXIgXyA9IHsgbGFiZWw6IDAsIHNlbnQ6IGZ1bmN0aW9uKCkgeyBpZiAodFswXSAmIDEpIHRocm93IHRbMV07IHJldHVybiB0WzFdOyB9LCB0cnlzOiBbXSwgb3BzOiBbXSB9LCBmLCB5LCB0LCBnO1xuICAgIHJldHVybiBnID0geyBuZXh0OiB2ZXJiKDApLCBcInRocm93XCI6IHZlcmIoMSksIFwicmV0dXJuXCI6IHZlcmIoMikgfSwgdHlwZW9mIFN5bWJvbCA9PT0gXCJmdW5jdGlvblwiICYmIChnW1N5bWJvbC5pdGVyYXRvcl0gPSBmdW5jdGlvbigpIHsgcmV0dXJuIHRoaXM7IH0pLCBnO1xuICAgIGZ1bmN0aW9uIHZlcmIobikgeyByZXR1cm4gZnVuY3Rpb24gKHYpIHsgcmV0dXJuIHN0ZXAoW24sIHZdKTsgfTsgfVxuICAgIGZ1bmN0aW9uIHN0ZXAob3ApIHtcbiAgICAgICAgaWYgKGYpIHRocm93IG5ldyBUeXBlRXJyb3IoXCJHZW5lcmF0b3IgaXMgYWxyZWFkeSBleGVjdXRpbmcuXCIpO1xuICAgICAgICB3aGlsZSAoZyAmJiAoZyA9IDAsIG9wWzBdICYmIChfID0gMCkpLCBfKSB0cnkge1xuICAgICAgICAgICAgaWYgKGYgPSAxLCB5ICYmICh0ID0gb3BbMF0gJiAyID8geVtcInJldHVyblwiXSA6IG9wWzBdID8geVtcInRocm93XCJdIHx8ICgodCA9IHlbXCJyZXR1cm5cIl0pICYmIHQuY2FsbCh5KSwgMCkgOiB5Lm5leHQpICYmICEodCA9IHQuY2FsbCh5LCBvcFsxXSkpLmRvbmUpIHJldHVybiB0O1xuICAgICAgICAgICAgaWYgKHkgPSAwLCB0KSBvcCA9IFtvcFswXSAmIDIsIHQudmFsdWVdO1xuICAgICAgICAgICAgc3dpdGNoIChvcFswXSkge1xuICAgICAgICAgICAgICAgIGNhc2UgMDogY2FzZSAxOiB0ID0gb3A7IGJyZWFrO1xuICAgICAgICAgICAgICAgIGNhc2UgNDogXy5sYWJlbCsrOyByZXR1cm4geyB2YWx1ZTogb3BbMV0sIGRvbmU6IGZhbHNlIH07XG4gICAgICAgICAgICAgICAgY2FzZSA1OiBfLmxhYmVsKys7IHkgPSBvcFsxXTsgb3AgPSBbMF07IGNvbnRpbnVlO1xuICAgICAgICAgICAgICAgIGNhc2UgNzogb3AgPSBfLm9wcy5wb3AoKTsgXy50cnlzLnBvcCgpOyBjb250aW51ZTtcbiAgICAgICAgICAgICAgICBkZWZhdWx0OlxuICAgICAgICAgICAgICAgICAgICBpZiAoISh0ID0gXy50cnlzLCB0ID0gdC5sZW5ndGggPiAwICYmIHRbdC5sZW5ndGggLSAxXSkgJiYgKG9wWzBdID09PSA2IHx8IG9wWzBdID09PSAyKSkgeyBfID0gMDsgY29udGludWU7IH1cbiAgICAgICAgICAgICAgICAgICAgaWYgKG9wWzBdID09PSAzICYmICghdCB8fCAob3BbMV0gPiB0WzBdICYmIG9wWzFdIDwgdFszXSkpKSB7IF8ubGFiZWwgPSBvcFsxXTsgYnJlYWs7IH1cbiAgICAgICAgICAgICAgICAgICAgaWYgKG9wWzBdID09PSA2ICYmIF8ubGFiZWwgPCB0WzFdKSB7IF8ubGFiZWwgPSB0WzFdOyB0ID0gb3A7IGJyZWFrOyB9XG4gICAgICAgICAgICAgICAgICAgIGlmICh0ICYmIF8ubGFiZWwgPCB0WzJdKSB7IF8ubGFiZWwgPSB0WzJdOyBfLm9wcy5wdXNoKG9wKTsgYnJlYWs7IH1cbiAgICAgICAgICAgICAgICAgICAgaWYgKHRbMl0pIF8ub3BzLnBvcCgpO1xuICAgICAgICAgICAgICAgICAgICBfLnRyeXMucG9wKCk7IGNvbnRpbnVlO1xuICAgICAgICAgICAgfVxuICAgICAgICAgICAgb3AgPSBib2R5LmNhbGwodGhpc0FyZywgXyk7XG4gICAgICAgIH0gY2F0Y2ggKGUpIHsgb3AgPSBbNiwgZV07IHkgPSAwOyB9IGZpbmFsbHkgeyBmID0gdCA9IDA7IH1cbiAgICAgICAgaWYgKG9wWzBdICYgNSkgdGhyb3cgb3BbMV07IHJldHVybiB7IHZhbHVlOiBvcFswXSA%2FIG9wWzFdIDogdm9pZCAwLCBkb25lOiB0cnVlIH07XG4gICAgfVxufTtcbmZ1bmN0aW9uIGZuKCkge1xuICAgIHJldHVybiBfX2F3YWl0ZXIodGhpcywgdm9pZCAwLCB2b2lkIDAsIGZ1bmN0aW9uICgpIHtcbiAgICAgICAgcmV0dXJuIF9fZ2VuZXJhdG9yKHRoaXMsIGZ1bmN0aW9uIChfYSkge1xuICAgICAgICAgICAgcmV0dXJuIFsyIC8qcmV0dXJuKi8sIHNsZWVwKCldO1xuICAgICAgICB9KTtcbiAgICB9KTtcbn1cbmZuKCkudGhlbihmdW5jdGlvbiAoKSB7IHJldHVybiBjb25zb2xlLmxvZygpOyB9KTsiLCJydW5zIjpbMTAyMDAwLDkzMDAwLDEwMDAsMjAzMDAwLDQxMTAwMCwzOTgwMDAsMTAwMCwyNTcwMDAsMzI2MDAwLDUxMTAwMCwxNjIwMDAsNDA1MDAwLDM0NjAwMCw1MTEwMDAsMzgzMDAwLDI5NzAwMCwzMDAwLDEwMDAsNDExMDAwLDQ4MTAwMCw1MTEwMDAsMTAwMDAsNTIyMDAwLDIxNzAwMCwxNTUwMDAsMTgyMDAwLDUxMTAwMCw3MzAwMCwxMzcwMDAsMTAwMCw0OTUwMDAsMjQzMDAwLDUxMTAwMCwzMDAwLDU1NjAwMCwxNjkwMDAsMTA3MDAwLDU5MjAwMCwyNTgwMDAsNDIyMDAwLDI2MTAwMCw0NjAwMCw1MTEwMDAsMTE5MDAwLDQ2MzAwMCw1NjcwMDAsMTM3MDAwLDQ0NjAwMCw3MDAwLDQ0MjAwMCwzODYwMDAsMjY4MDAwLDEwMDAsNTExMDAwLDEwMDAsMTAwMCw1NDAwMCwxMDAwLDQyNjAwMCwxODQwMDAsMjM4MDAwLDY4NTAwMCw0NjAwMDAsMTAwMCw0NDYwMDAsMTAwMCwzNTAwMDAsMzM1MDAwLDQxMTAwMCwxNzIwMDAsMTM3MDAwLDEzNzAwMCwzNTcwMDAsMjA0MDAwLDM3MDAwLDM1MDAwMCw0MTEwMDAsMTM3MDAwLDM4NzAwMCwzNjUwMDAsODMwMDAsMTUwMDAwLDE2MDAwLDcwMDAwLDM3NjAwMCwxMDAwLDYzMzAwMCw1MDkwMDAsNDM0MDAwLDE2MDAwLDUzNjAwMCwxMzcwMDAsNTM3MDAwLDQ5MDAwLDgwMDAsNDEwMDAwLDEzNzAwMCwyNTQwMDAsNDY0MDAwLDEzNzAwMF0sIm9wcyI6MjYzODgwfSx7Im5hbWUiOiJBc3luYyBBd2FpdCIsImNvZGUiOiJhc3luYyBmdW5jdGlvbiBmbigpIHtcbiAgcmV0dXJuIGF3YWl0IHNsZWVwKCk7XG59XG5mbigpLnRoZW4oKCkgPT4gY29uc29sZS5sb2coKSk7IiwicnVucyI6WzI3MTcwMDAsMTM1NTAwMCw3ODAwMDAsMjcxNzAwMCwxODAwMCwxMDI1MDAwLDkyMTAwMCwyMjI4MDAwLDIyMjgwMDAsMjcxNzAwMCwxMDU1MDAwLDE4MTYwMDAsMjcxNzAwMCwyMTAwMCwyNDg1MDAwLDI2MDEwMDAsMjcxNzAwMCw4MTcwMDAsNDkwMDAsNzAwMCwyNzE3MDAwLDUxMTAwMCwyNzE3MDAwLDE1NzgwMDAsMjEwMDAwLDIwODkwMDAsMjcxNzAwMCwxMTQzMDAwLDI3MTcwMDAsOTAwMCwyNTY2MDAwLDE4OTYwMDAsMzM1MDAwLDU3MjAwMCwyNzE3MDAwLDgyMTAwMCwyNzE3MDAwLDI3MTcwMDAsMTMzNDAwMCwyNDM3MDAwLDI1NjkwMDAsODczMDAwLDI3MTcwMDAsMjM1NjAwMCwyMTEyMDAwLDI3MTcwMDAsMTI4NDAwMCwyNTU0MDAwLDkzNTAwMCwxMDAwLDEyODMwMDAsMTgyMjAwMCwyNzE3MDAwLDIwMzQwMDAsNDM4MDAwLDI5MzAwMCwxMjU0MDAwLDU1MTAwMCwyNzE3MDAwLDE4NTgwMDAsMTQwNDAwMCwyNzE3MDAwLDI2NDIwMDAsNTk4MDAwLDI3MTcwMDAsNDQwMDAsMjcxNzAwMCwyMzkxMDAwLDQwNTAwMCwxOTk4MDAwLDU1NjAwMCwyMDAwLDE5NDYwMDAsNjQ4MDAwLDgzNDAwMCwyNzE3MDAwLDI3MTcwMDAsODQ3MDAwLDI3MTcwMDAsMjAzMTAwMCw1MTEwMDAsMTkwNjAwMCw5NzQwMDAsNTExMDAwLDIzOTMwMDAsMjcxNzAwMCwxNjEyMDAwLDk1MDAwLDI3MTcwMDAsNjI1MDAwLDEyMjYwMDAsNjEzMDAwLDI3MTcwMDAsMTE5NTAwMCw3MDAwMDAsNTY3MDAwLDQ5OTAwMCwxMjA5MDAwLDIzNjEwMDAsMTMxNTAwMF0sIm9wcyI6MTU3NzA3MH0seyJuYW1lIjoiQXN5bmMgQXdhaXQgRG93bmxldmVsbGVkIiwiY29kZSI6InZhciBfX2F3YWl0ZXIgPSAodGhpcyAmJiB0aGlzLl9fYXdhaXRlcikgfHwgZnVuY3Rpb24gKHRoaXNBcmcsIF9hcmd1bWVudHMsIFAsIGdlbmVyYXRvcikge1xuICAgIGZ1bmN0aW9uIGFkb3B0KHZhbHVlKSB7IHJldHVybiB2YWx1ZSBpbnN0YW5jZW9mIFAgPyB2YWx1ZSA6IG5ldyBQKGZ1bmN0aW9uIChyZXNvbHZlKSB7IHJlc29sdmUodmFsdWUpOyB9KTsgfVxuICAgIHJldHVybiBuZXcgKFAgfHwgKFAgPSBQcm9taXNlKSkoZnVuY3Rpb24gKHJlc29sdmUsIHJlamVjdCkge1xuICAgICAgICBmdW5jdGlvbiBmdWxmaWxsZWQodmFsdWUpIHsgdHJ5IHsgc3RlcChnZW5lcmF0b3IubmV4dCh2YWx1ZSkpOyB9IGNhdGNoIChlKSB7IHJlamVjdChlKTsgfSB9XG4gICAgICAgIGZ1bmN0aW9uIHJlamVjdGVkKHZhbHVlKSB7IHRyeSB7IHN0ZXAoZ2VuZXJhdG9yW1widGhyb3dcIl0odmFsdWUpKTsgfSBjYXRjaCAoZSkgeyByZWplY3QoZSk7IH0gfVxuICAgICAgICBmdW5jdGlvbiBzdGVwKHJlc3VsdCkgeyByZXN1bHQuZG9uZSA%2FIHJlc29sdmUocmVzdWx0LnZhbHVlKSA6IGFkb3B0KHJlc3VsdC52YWx1ZSkudGhlbihmdWxmaWxsZWQsIHJlamVjdGVkKTsgfVxuICAgICAgICBzdGVwKChnZW5lcmF0b3IgPSBnZW5lcmF0b3IuYXBwbHkodGhpc0FyZywgX2FyZ3VtZW50cyB8fCBbXSkpLm5leHQoKSk7XG4gICAgfSk7XG59O1xudmFyIF9fZ2VuZXJhdG9yID0gKHRoaXMgJiYgdGhpcy5fX2dlbmVyYXRvcikgfHwgZnVuY3Rpb24gKHRoaXNBcmcsIGJvZHkpIHtcbiAgICB2YXIgXyA9IHsgbGFiZWw6IDAsIHNlbnQ6IGZ1bmN0aW9uKCkgeyBpZiAodFswXSAmIDEpIHRocm93IHRbMV07IHJldHVybiB0WzFdOyB9LCB0cnlzOiBbXSwgb3BzOiBbXSB9LCBmLCB5LCB0LCBnO1xuICAgIHJldHVybiBnID0geyBuZXh0OiB2ZXJiKDApLCBcInRocm93XCI6IHZlcmIoMSksIFwicmV0dXJuXCI6IHZlcmIoMikgfSwgdHlwZW9mIFN5bWJvbCA9PT0gXCJmdW5jdGlvblwiICYmIChnW1N5bWJvbC5pdGVyYXRvcl0gPSBmdW5jdGlvbigpIHsgcmV0dXJuIHRoaXM7IH0pLCBnO1xuICAgIGZ1bmN0aW9uIHZlcmIobikgeyByZXR1cm4gZnVuY3Rpb24gKHYpIHsgcmV0dXJuIHN0ZXAoW24sIHZdKTsgfTsgfVxuICAgIGZ1bmN0aW9uIHN0ZXAob3ApIHtcbiAgICAgICAgaWYgKGYpIHRocm93IG5ldyBUeXBlRXJyb3IoXCJHZW5lcmF0b3IgaXMgYWxyZWFkeSBleGVjdXRpbmcuXCIpO1xuICAgICAgICB3aGlsZSAoZyAmJiAoZyA9IDAsIG9wWzBdICYmIChfID0gMCkpLCBfKSB0cnkge1xuICAgICAgICAgICAgaWYgKGYgPSAxLCB5ICYmICh0ID0gb3BbMF0gJiAyID8geVtcInJldHVyblwiXSA6IG9wWzBdID8geVtcInRocm93XCJdIHx8ICgodCA9IHlbXCJyZXR1cm5cIl0pICYmIHQuY2FsbCh5KSwgMCkgOiB5Lm5leHQpICYmICEodCA9IHQuY2FsbCh5LCBvcFsxXSkpLmRvbmUpIHJldHVybiB0O1xuICAgICAgICAgICAgaWYgKHkgPSAwLCB0KSBvcCA9IFtvcFswXSAmIDIsIHQudmFsdWVdO1xuICAgICAgICAgICAgc3dpdGNoIChvcFswXSkge1xuICAgICAgICAgICAgICAgIGNhc2UgMDogY2FzZSAxOiB0ID0gb3A7IGJyZWFrO1xuICAgICAgICAgICAgICAgIGNhc2UgNDogXy5sYWJlbCsrOyByZXR1cm4geyB2YWx1ZTogb3BbMV0sIGRvbmU6IGZhbHNlIH07XG4gICAgICAgICAgICAgICAgY2FzZSA1OiBfLmxhYmVsKys7IHkgPSBvcFsxXTsgb3AgPSBbMF07IGNvbnRpbnVlO1xuICAgICAgICAgICAgICAgIGNhc2UgNzogb3AgPSBfLm9wcy5wb3AoKTsgXy50cnlzLnBvcCgpOyBjb250aW51ZTtcbiAgICAgICAgICAgICAgICBkZWZhdWx0OlxuICAgICAgICAgICAgICAgICAgICBpZiAoISh0ID0gXy50cnlzLCB0ID0gdC5sZW5ndGggPiAwICYmIHRbdC5sZW5ndGggLSAxXSkgJiYgKG9wWzBdID09PSA2IHx8IG9wWzBdID09PSAyKSkgeyBfID0gMDsgY29udGludWU7IH1cbiAgICAgICAgICAgICAgICAgICAgaWYgKG9wWzBdID09PSAzICYmICghdCB8fCAob3BbMV0gPiB0WzBdICYmIG9wWzFdIDwgdFszXSkpKSB7IF8ubGFiZWwgPSBvcFsxXTsgYnJlYWs7IH1cbiAgICAgICAgICAgICAgICAgICAgaWYgKG9wWzBdID09PSA2ICYmIF8ubGFiZWwgPCB0WzFdKSB7IF8ubGFiZWwgPSB0WzFdOyB0ID0gb3A7IGJyZWFrOyB9XG4gICAgICAgICAgICAgICAgICAgIGlmICh0ICYmIF8ubGFiZWwgPCB0WzJdKSB7IF8ubGFiZWwgPSB0WzJdOyBfLm9wcy5wdXNoKG9wKTsgYnJlYWs7IH1cbiAgICAgICAgICAgICAgICAgICAgaWYgKHRbMl0pIF8ub3BzLnBvcCgpO1xuICAgICAgICAgICAgICAgICAgICBfLnRyeXMucG9wKCk7IGNvbnRpbnVlO1xuICAgICAgICAgICAgfVxuICAgICAgICAgICAgb3AgPSBib2R5LmNhbGwodGhpc0FyZywgXyk7XG4gICAgICAgIH0gY2F0Y2ggKGUpIHsgb3AgPSBbNiwgZV07IHkgPSAwOyB9IGZpbmFsbHkgeyBmID0gdCA9IDA7IH1cbiAgICAgICAgaWYgKG9wWzBdICYgNSkgdGhyb3cgb3BbMV07IHJldHVybiB7IHZhbHVlOiBvcFswXSA%2FIG9wWzFdIDogdm9pZCAwLCBkb25lOiB0cnVlIH07XG4gICAgfVxufTtcbmZ1bmN0aW9uIGZuKCkge1xuICAgIHJldHVybiBfX2F3YWl0ZXIodGhpcywgdm9pZCAwLCB2b2lkIDAsIGZ1bmN0aW9uICgpIHtcbiAgICAgICAgcmV0dXJuIF9fZ2VuZXJhdG9yKHRoaXMsIGZ1bmN0aW9uIChfYSkge1xuICAgICAgICAgICAgc3dpdGNoIChfYS5sYWJlbCkge1xuICAgICAgICAgICAgICAgIGNhc2UgMDogcmV0dXJuIFs0IC8qeWllbGQqLywgc2xlZXAoKV07XG4gICAgICAgICAgICAgICAgY2FzZSAxOiByZXR1cm4gWzIgLypyZXR1cm4qLywgX2Euc2VudCgpXTtcbiAgICAgICAgICAgIH1cbiAgICAgICAgfSk7XG4gICAgfSk7XG59XG5mbigpLnRoZW4oZnVuY3Rpb24gKCkgeyByZXR1cm4gY29uc29sZS5sb2coKTsgfSk7XG4iLCJydW5zIjpbMTM3MDAwLDE4NTAwMCwxMDAwLDUzMzAwMCw0MDYwMDAsMTM3MDAwLDEwMDAsMTAwMCw1MzAwMCw1MTEwMDAsMTAwMCw0MTEwMDAsMzc5MDAwLDUxMTAwMCwxNzkwMDAsNzMwMDAsMTE1MDAwLDEwMDAsNTA3MDAwLDQxMTAwMCwyNjcwMDAsMzU2MDAwLDM2MjAwMCw1NzYwMDAsMTczMDAwLDE2NDAwMCw1MTEwMDAsNzAwMDAsMjcwMDAwLDI4MTAwMCwyMzAwMCwxMzcwMDAsNDg5MDAwLDEwMDAsNTAwMDAsMTAwMCw1OTYwMDAsNDExMDAwLDEwMDAsNDQzMDAwLDI1MDAwLDExMDAwLDQ5ODAwMCw0NDEwMDAsMTQ2MDAwLDE0ODAwMCwxNjQwMDAsMTAwMDAwLDQwMDAsNTM4MDAwLDEwMDAsNjYwMDAsMzc5MDAwLDE1NzAwMCw1MDEwMDAsNDUyMDAwLDEwMDAsMTAwMCwxNTMwMDAsNDIyMDAwLDMyNTAwMCw0MjIwMDAsMTAwMCw0MzQwMDAsMzYyMDAwLDMxODAwMCw4ODAwMCwzMDYwMDAsNTExMDAwLDEzNzAwMCw1MTMwMDAsMjcwMDAwLDQxMTAwMCw0MzgwMDAsMzEwMDAsMzg2MDAwLDM1MjAwMCw3NjcwMDAsMTM3MDAwLDIxMDAwLDE1OTAwMCwxMjkwMDAsODAwMCwzMDcwMDAsNjQwMDAsMzAyMDAwLDE4MDAwLDQ5NjAwMCw0NDgwMDAsMjIwMDAsMzU5MDAwLDMyNjAwMCwxMzcwMDAsMjAwMCwxMDAwLDEwMDAsNTU5MDAwLDY2MDAwLDE4ODAwMCwxMzcwMDBdLCJvcHMiOjIzOTAyMH1dLCJ1cGRhdGVkIjoiMjAyNC0wNi0wNlQwNDoyMDoxMS43MzdaIn0%3D)

- no async = ~3.2m ops/s
- with async = ~2m ops/s
- with async + await = ~1.5m ops/s

if you're downlevelling:
- with async = ~400k ops/s
- with async + await = ~300k ops/s

Which illustrates the perf difference. However that difference is - as I mentioned - nanoseconds. You lose 1.2m ops/s by having an unnecessary await. Which means that the unnecessary await is ~1 nanosecond slower.

Unless you're downlevelling to <ES2017 - if you're doing that then the polyfilled async/await is much slower.
